streams/src/main/java/org/apache/kafka/streams/processor/internals/ProcessorStateManager.java

```diff
@@ -129,50 +129,52 @@ public void register(StateStore store, boolean loggingEnabled, StateRestoreCallb
         } else if (sourceStoreToSourceTopic != null && sourceStoreToSourceTopic.containsKey(store.name())) {
             topic = sourceStoreToSourceTopic.get(store.name());
-        } else {
-            throw new IllegalArgumentException("Store is neither built from source topic, nor has a changelog.");
         }
 
-        // block until the partition is ready for this state changelog topic or time has elapsed
-        int partition = getPartition(topic);
-        boolean partitionNotFound = true;
-        long startTime = System.currentTimeMillis();
-        long waitTime = 5000L;      // hard-code the value since we should not block after KIP-4
-
-        do {
-            try {
-                Thread.sleep(50L);
-            } catch (InterruptedException e) {
-                // ignore
-            }
+        if (topic != null) {
+            // block until the partition is ready for this state changelog topic or time has elapsed
+            int partition = getPartition(topic);
+            boolean partitionNotFound = true;
+            long startTime = System.currentTimeMillis();
+            long waitTime = 5000L;      // hard-code the value since we should not block after KIP-4
+
+            do {
+                try {
+                    Thread.sleep(50L);
+                } catch (InterruptedException e) {
+                    // ignore
+                }
 
-            List<PartitionInfo> partitionInfos = restoreConsumer.partitionsFor(topic);
-            if (partitionInfos == null) {
-                throw new StreamsException("Could not find partition info for topic: " + topic);
-            }
-            for (PartitionInfo partitionInfo : partitionInfos) {
-                if (partitionInfo.partition() == partition) {
-                    partitionNotFound = false;
-                    break;
+                List<PartitionInfo> partitionInfos = restoreConsumer.partitionsFor(topic);
+                if (partitionInfos == null) {
+                    throw new StreamsException("Could not find partition info for topic: " + topic);
                 }
-            }
-        } while (partitionNotFound && System.currentTimeMillis() < startTime + waitTime);
+                for (PartitionInfo partitionInfo : partitionInfos) {
+                    if (partitionInfo.partition() == partition) {
+                        partitionNotFound = false;
+                        break;
+                    }
+                }
+            } while (partitionNotFound && System.currentTimeMillis() < startTime + waitTime);
 
-        if (partitionNotFound)
-            throw new StreamsException("Store " + store.name() + "'s change log (" + topic + ") does not contain partition " + partition);
+            if (partitionNotFound)
+                throw new StreamsException("Store " + store.name() + "'s change log (" + topic + ") does not contain partition " + partition);
+        }
 
         this.stores.put(store.name(), store);
```

Comment:
I think this should only be done after the store is in a valid state, i.e, after restore. Otherwise there is a chance we can try and query or write to the store before it is ready